File: problems/Construct Binary Search Tree from Preorder Traversal/bstFromPreorder.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode BstFromPreorder(int[] preorder) {
        if (null == preorder || 0 == preorder.Length) {
            return null;
        }

        var root = new TreeNode(preorder[0]);
        var stack = new Stack<TreeNode>();

        stack.Push(root);

        for (var i = 1; preorder.Length > i; ++i) {
            var parent = stack.Peek();
            var child = new TreeNode(preorder[i]);

            while (0 < stack.Count && stack.Peek().val < child.val) {
                parent = stack.Pop();
            }

            if (parent.val < child.val) {
                parent.right = child;
            } else {
                parent.left = child;
            }

            stack.Push(child);
        }

        return root;
    }
}

File: problems/Construct Binary Search Tree from Preorder Traversal/bstFromPreorder.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode bstFromPreorder(int[] preorder) {
        int n = preorder.length;

        if (0 == n) {
            return null;
        }

        TreeNode root = new TreeNode(preorder[0]);
        Stack<TreeNode> stack = new Stack<TreeNode>();

        stack.push(root);

        for (int idx = 1; n > idx; ++idx) {
            TreeNode node = stack.peek();
            TreeNode child = new TreeNode(preorder[idx]);

            while (!stack.isEmpty() && stack.peek().val < child.val) {
                node = stack.pop();
            }

            if (node.val < child.val) {
                node.right = child;
            } else {
                node.left = child;
            }

            stack.push(child);
        }

        return root;
    }
}
